mycompany-seeq-link-connector-myconnector/src/main/java/com/mycompany/seeq/link/connector/DatasourceSimulator.java

```diff
@@ -30,55 +94,177 @@ public String getName() {
-        this.tagCount = tagCount;
+    // To be able to yield consistent, reproducible tag values, we need a constant seed. This helps us
+    // approximate the behaviour of a real datasource which should be deterministic.
+    private final int RandomnessSeed = 1_000_000;
+
+    private final Random RNG = new Random(RandomnessSeed);
+
+    private boolean connected;
+    private Duration signalPeriod;
+
+    public DatasourceSimulator(Duration signalPeriod) {
         this.signalPeriod = signalPeriod;
     }
 
+    public boolean isConnected() {
+        return this.connected;
+    }
+
     public boolean connect() {
         this.connected = true;
 
         return true;
     }
 
-    public boolean isConnected() {
-        return this.connected;
+    public void disconnect() {}
+
+    public Iterable<Element> getDatabases() {
+        int databaseCount = RNG.nextInt(10);
+        return IntStream.range(1, databaseCount + 1)
+                .mapToObj(Element::new)
+                .collect(Collectors.toList());
     }
 
-    public void disconnect() {}
+    public Iterable<Alarm> getAlarmsForDatabase(String elementId) {
+        int alarmCount = RNG.nextInt(10);
+        return IntStream.range(1, alarmCount + 1)
+                .mapToObj(alarmId -> new Alarm(elementId, alarmId))
+                .collect(Collectors.toList());
+    }
+
+    public Iterable<Tag> getTagsForDatabase(String elementId) {
+        int tagCount = RNG.nextInt(10);
+        return IntStream.range(1, tagCount + 1)
+                .mapToObj(tagId -> new Tag(elementId, tagId, tagId % 2 == 0))
+                .collect(Collectors.toList());
+    }
+
+    public Iterable<Constant> getConstantsForDatabase(String elementId) {
+        int constantCount = RNG.nextInt(10);
+        return IntStream.range(1, constantCount + 1)
+                .mapToObj(constId -> new Constant(elementId, constId, "°C", constId * 10))
+                .collect(Collectors.toList());
+    }
 
-    public int getTagCount() {
-        return this.tagCount;
+    public enum Waveform {
+        SINE
     }
 
-    public Iterator<Tag> getTags() {
-        return new Iterator<Tag>() {
-            private int nextTagNumber = 0;
+    public Iterable<Tag.Value> getTagValues(String dataId, TimeInstant startTimestamp, TimeInstant endTimestamp,
+            int limit) {
+        long samplePeriodInNanos = this.signalPeriod.toNanos();
+        long leftBoundTimestamp = startTimestamp.getTimestamp() / samplePeriodInNanos;
+        long rightBoundTimestamp = (endTimestamp.getTimestamp() + samplePeriodInNanos - 1) / samplePeriodInNanos;
+
+        return () -> new Iterator<>() {
+            private long sampleIndex = leftBoundTimestamp;
+            private long count = 0;
 
             @Override
             public boolean hasNext() {
-                return (this.nextTagNumber < DatasourceSimulator.this.tagCount);
+                return sampleIndex <= rightBoundTimestamp && count < limit;
```

Comment:
Question: shouldn't the first condition be `sampleIndex * samplePeriodInNanos <= rightBoundTimestamp`?